Add support for @nullable trait in order to support MergePatch usecases

astridej opened the pull request:
This PR is an update of https://github.com/disneystreaming/smithy4s/pull/1151 to take into account the changes to default and required handling that have happened since and to also change the modelling a little in line with what was discussed in the Discord (in particular, modelling required vs nullable as independent of each other). The change was big enough that I ended up writing the code from scratch, but I definitely based it on the linked PR.

With the change, it is possible to mark a field as nullable in order to distinguish between absence and removal of a field. This is modelled in Scala via `Option[Nullable[A]]` - `Nullable` being isomorphic to `Option` and left as a separate class for clarity. Unfortunately, this opens up a can of worms regarding possible combinations with @required and @default, the vast majority of which make no sense but are still theoretically possible. I had to rework `Field.Modifier` to make sure all possible combinations were reflected and handled properly, and although I did my best to keep the resulting complexity from leaking out it does show up in some places 😬

I think I've now got this in roughly the state the original PR was in, so opening a draft PR in order to get feedback while I work on the missing pieces.

## PR Checklist (not all items are relevant to all PRs)
    
- [✅] Added unit-tests (for runtime code)
- [✅] Added bootstrapped code + smoke tests (when the rendering logic is modified)
- [👎] Added build-plugins integration tests (when reflection loading is required at codegen-time)
- [👎] Added alloy compliance tests (when simpleRestJson protocol behaviour is expanded/updated)
- [✅ ] Updated dynamic module to match generated-code behaviour
- [✅ ] Added documentation
- [✅ ] Updated changelog

astridej committed:
Add nullable support to JSON parsing

The JSON codec will now check for a Nullable hint on the schema it is
parsing and not immediately parse fields as None if present. Note that
getting this to work with a new Nullable type does require adjusting
the code checking whether a Schema is for an Option a little so that it
can also include bijections from an option type.

astridej committed:
Adjust codegen code to handle Nullable

This involves totally reworking the Field.Modifier logic, as the
addition of Nullable makes the number of cases too complex to keep track
of in an enum alone: the modifier logic now depends on
* whether there is a @required annotation present
* whether there is a @nullable annotation present
* whether there is a default value present, and (for nullable) whether
this value is null or a concrete value

making for a total of ten theoretical cases resulting in four possible
types in the Scala code. We model this in a case class and attempt to
shift some of the logic, such as ordering or working out the wrapped
type, to the class itself in order to avoid other parts of the code
needing to work with its internals.

astridej committed:
Add nullable sample spec and smoke tests

There is a mild tension here: on the one hand, we really want to test
all the edge cases of different field modifier combinations to make sure
the generation code works as wel expect. On the other, these combinations
generally have only an extremely limited use or are just overly convoluted
ways to specify the default behaviour, and so they shouldn't be in the
sample specs people use to check usage. We handle this by providing two
test cases: Patchable, which handles the merge patch behaviour this whole
change is meant to support, and PatchableEdgeCases which is explicitly
described as only there for testing purposes.

astridej committed:
Inexplicable hint reorderings that happened on compilation

???

astridej committed:
Minor refactor

astridej committed:
Scalafmt

astridej committed:
Rework get message to be safer until I can test it

astridej committed:
Remove commented out code + cleanup

astridej commented on modules/bootstrapped/src/generated/com/amazonaws/dynamodb/Endpoint.scala:
```diff
@@ -25,7 +25,7 @@ object Endpoint extends ShapeTag.Companion[Endpoint] {
   implicit val schema: Schema[Endpoint] = struct(
     string.required[Endpoint]("Address", _.address).addHints(smithy.api.Documentation("<p>IP address of the endpoint.</p>")),
-    long.required[Endpoint]("CachePeriodInMinutes", _.cachePeriodInMinutes).addHints(smithy.api.Default(smithy4s.Document.fromDouble(0.0d)), smithy.api.Documentation("<p>Endpoint cache time to live (TTL) value.</p>")),
+    long.required[Endpoint]("CachePeriodInMinutes", _.cachePeriodInMinutes).addHints(smithy.api.Documentation("<p>Endpoint cache time to live (TTL) value.</p>"), smithy.api.Default(smithy4s.Document.fromDouble(0.0d))),
```
I'm not exactly sure why the hint ordering changed on some of the generated code or whether that's a problem? If it is, I can dig into it.

Baccata replied:
mmm it may depend on non-deterministic stuff (ordering of map and sets). Not a problem unless the CI starts shouting (which it might), in which case we'll force a ShapeId-based ordering or something.

astridej commented on modules/codegen/src/smithy4s/codegen/internals/IR.scala:
```diff
@@ -141,29 +142,51 @@ private[internals] case class StreamingField(
+      nullable: Boolean,
+      default: Option[Default]
+  ) {
+    def typeMod: TypeModification =
```
This case class is unfortunately complex, so I tried to make sure most places using it didn't actually need to destructure it. The `typeMod` field is part of that, since a bunch of consumers of this class only care about whether the resulting type is wrapped in an `Option` (or `Nullable`, or `Option[Nullable[...]]`).

I did look into converting the underlying type representation to a `Nullable` before reaching the point where the modifier was generated - in which case in theory nothing here should need to change - but I was scared off by the fact that nothing in that part of the code seemed to be doing modifications of the underlying shape's type. I can have a second look if people think it's better there, though - I'm still not happy with how complex the modifier logic is now. 🤷‍♀️

Baccata replied:
I'll give it a think, but worst case scenario I'm not too bothered : the internal representation is implementation detail, and even though the complexity does increase with the "nullable" requirement, it's still acceptable imho

Baccata replied:
I tried to think of alternative ways of modelling this but what potentially made the code look nicer (using composable binary flags, for instance) or custom extractors also lose us exhaustive pat-mat, so ... I'm fine with the seemingly ugliness if you can't find a better way of modelling it.

astridej replied:
Yeah, that was pretty much my conclusion too 😓 . At least `typeMod` got rid of a bunch of the places that would have needed to destructure this case class otherwise.

astridej commented on modules/codegen/src/smithy4s/codegen/internals/Renderer.scala:
```diff
@@ -811,17 +808,18 @@ private[internals] class Renderer(compilationUnit: CompilationUnit) { self =>
-        line"override def getMessage(): $string_ = ${field.name}.value"
-      case field =>
-        line"override def getMessage(): $string_ = ${field.name}.map(_.value).orNull"
+    val fetchLogic = field.modifier.typeMod match {
```
I still need to test that this works the way it should. (Even if part of me wants to say that anyone who annotates a field with `@errorMessage @nullable @required @default` or the like is getting what they deserve 🙈 )

Baccata replied:
😆 yeah unfortunately this project is very much subject to Murphy's law, so I'm thankful with the additional patterns

astridej replied:
😂 OK, added another detailed smoke spec test that goes through the eight possible cases and makes sure it can retrieve the error message.

astridej commented on modules/codegen/src/smithy4s/codegen/internals/SmithyToIR.scala:
```diff
@@ -1239,17 +1230,19 @@ private[codegen] class SmithyToIR(model: Model, namespace: String) {
+            } // value or default must be present if type is not wrapped
+            TypedNode.FieldTN.RequiredTN(NodeAndType(node, tpe))
+          case Field(_, realName, tpe, _, _) =>
+            // TODO AJ: anything else need to happen here?
```
See todo comment - given the extra complexity involving `Nullable`, the current set of `FieldTN` values doesn't seem to accurately reflect the state of the field. That said, everything _seems_ to be working fine and this part of the code wasn't touched further in the original PR, so I'm not sure if I need to update them.

Baccata replied:
I think you're not gonna need to do anything else : this code is aimed at rendering smithy trait as smithy4s hints. Smithy traits are expressed using a "canonical" node representation, which doesn't take into consideration any other trait but the `@required` trait. 

This kinda implies that we should probably protect against the application of the `@nullable` trait on shapes that are traits (or used in traits). 

You can avoid worrying about it.

astridej commented on modules/core/src/smithy4s/schema/Schema.scala:
```diff
@@ -90,8 +90,11 @@ sealed trait Schema[A]{
+
   final def isOption: Boolean = this match {
     case _: OptionSchema[_] => true
+    case BijectionSchema(underlying, _) => underlying.isOption // TODO AJ: does this have any side-effects in the other places .isOption is used? is it a breaking change?
```
See TODO comment...

This is necessary in order to get the new code in `DocumentDecoderSchemaVisitor` and `SchemaVisitorJCodec` (which I copied from the original PR) to work for `Nullable.schema`, but I'm not sure if it breaks anything else. The main alternatives I can see are to drop that check and just always delegate null values to the underlying decoder if a `nullable` hint is present (is that safe?), or define the new schema here instead of as a bijection from option.

Baccata commented on modules/bootstrapped/src/generated/smithy4s/example/PatchableEdgeCases.scala:
```diff
@@ -0,0 +1,28 @@
+import smithy4s.schema.Schema.int
+import smithy4s.schema.Schema.struct
+
+final case class PatchableEdgeCases(required: Nullable[Int], requiredDefaultValue: Nullable[Int] = smithy4s.Nullable.Value(3), requiredDefaultNull: Nullable[Int] = Null, defaultValue: Nullable[Int] = smithy4s.Nullable.Value(5), defaultNull: Nullable[Int] = Null)
```
this is great, thank you !

Baccata commented on modules/core/src/smithy4s/schema/Schema.scala:
```diff
@@ -90,8 +90,11 @@ sealed trait Schema[A]{
+
   final def isOption: Boolean = this match {
     case _: OptionSchema[_] => true
+    case BijectionSchema(underlying, _) => underlying.isOption // TODO AJ: does this have any side-effects in the other places .isOption is used? is it a breaking change?
```
I don't think `.isOption` is used anymore. It was probably used transiently, but I think the change is fine. 

Probably worth adding a `case RefinementSchema` too, just in case.

astridej replied:
It shows up once in a place involving `flexibleCollectionSupport` in the JSON codec ( https://github.com/disneystreaming/smithy4s/blob/aaf8b1f034b7d3f97e6f250c97fa08c517686d28/modules/json/src/smithy4s/json/internals/SchemaVisitorJCodec.scala#L895 ) but I'm not sure what the logic there is doing.

And good point on the RefinementSchema - will add that!

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/signed)](https://cla-assistant.io/disneystreaming/smithy4s?pullRequest=1408) <br/>All committers have signed the CLA.

astridej committed:
Add refinement schema to expanded .isOption logic

Baccata commented on modules/core/src/smithy4s/Nullable.scala:
```diff
@@ -0,0 +1,62 @@
+    case Null     => Null
+  }
+
+  def fold[B](whenValue: A => B, whenNull: => B): B =
```
I think it'd be good to match the signature of Option's fold : 

```
def fold[B](whenNull : => B)(whenValue: A => B) : B
```

astridej replied:
Shall do!

Also, just realising - would it make sense to put some type classes for `Nullable` in the `cats` module?

Baccata replied:
I suppose it would 👍

Baccata commented on modules/core/src/smithy4s/Nullable.scala:
```diff
@@ -0,0 +1,62 @@
+    case None    => Null
+  }
+
+  case class Value[A](a: A) extends Nullable[A]
```
```suggestion
  final case class Value[A](a: A) extends Nullable[A]
```

Baccata commented:
@astridej, one important aspect of this work is gonna be the addition of tests to [this specification](https://github.com/disneystreaming/smithy4s/blob/aaf8b1f034b7d3f97e6f250c97fa08c517686d28/modules/bootstrapped/test/src/smithy4s/schema/FieldSpec.scala#L9). 

Most encoders now use the `foreachUnlessDefault` or `getUnlessDefault`, and we need to ensure that those behave in a somewhat coherent manner with respect to the `nullable` trait. 

I presume that this line may have to change in order to account for the `nullable` trait  

https://github.com/disneystreaming/smithy4s/blob/aaf8b1f034b7d3f97e6f250c97fa08c517686d28/modules/core/src/smithy4s/schema/DefaultValueSchemaVisitor.scala#L97

astridej committed:
Update modules/core/src/smithy4s/Nullable.scala

Co-authored-by: Olivier Mélois <baccata64@gmail.com>

astridej committed:
Add tests for error messages

The error message rendering code for custom error message fields is
different for every type modification and also for custom vs primitive
types - we add a test case for every combination to be absolutely sure
they're being rendered correctly.

astridej committed:
Revert "Inexplicable hint reorderings that happened on compilation"

This reverts commit 8edf43091141973b5a66969876e1f8376ce223a4.

astridej committed:
Use Option ordering for Nullable fold

astridej committed:
Remove unneeded TODO comment

Although technically the FieldTN representations aren't correct,
this part of the code should never use the @nullable annotation and
it's considered acceptable to leave as is.

astridej commented:
> @astridej, one important aspect of this work is gonna be the addition of tests to [this specification](https://github.com/disneystreaming/smithy4s/blob/aaf8b1f034b7d3f97e6f250c97fa08c517686d28/modules/bootstrapped/test/src/smithy4s/schema/FieldSpec.scala#L9).
> 
> Most encoders now use the `foreachUnlessDefault` or `getUnlessDefault`, and we need to ensure that those behave in a somewhat coherent manner with respect to the `nullable` trait.
> 
> I presume that this line may have to change in order to account for the `nullable` trait
> 
> https://github.com/disneystreaming/smithy4s/blob/aaf8b1f034b7d3f97e6f250c97fa08c517686d28/modules/core/src/smithy4s/schema/DefaultValueSchemaVisitor.scala#L97

@Baccata  I just had a look at this and wrote some tests. I can push them, but I should note that it looks like with the current implementation, the code is actually totally independent of `nullable` and doesn't have to change at all 😄. Digging into it, it seems to come down to two things:

* in this part of the code, `nullable` is handled in a different place from required/default - we first transform the schema into a nullable schema which sets the nullable hint, and then we call `.optional`, `.required`, or `.field` depending on the required/default situation to create the `Field` from it. `getUnlessDefault` can then ignore the nullable trait in the same it can ignore any other type refinement or complex underlying schema for a field. The only thing that could go wrong is that it could screw up on calculating the default, but...
* although there's no special logic for handling `Document.DNull` in the nullable case here, it does the right thing anyway: on receiving null, the logic gets forwarded to `DefaultValueSchemaVisitor`, which attempts to calculate the default value for a schema for `Nullable[T]` and ends up with `Nullable.Null` via the bijection to `Option`, which is what we would've wanted anyway.

If I'm off-base here, let me know!

astridej committed:
Add cats instances for Nullable

This does not include every type class satisfied by Option but just the
ones we think are most likely to be needed by a consumer.

astridej committed:
Add test and refactor for getUnlessDefault

We refactor foreachUnlessDefault so the logic is not duplicated across
two methods, and add some tests to confirm that it works as expected on
a nullable field.

astridej committed:
Scalafmt

Baccata commented on modules/core/src/smithy4s/schema/Field.scala:
```diff
@@ -65,10 +65,8 @@ final case class Field[S, A](
    */
-  def foreachUnlessDefault(s: S)(f: A => Unit): Unit = {
-    val a = get(s)
-    if (isDefaultValue(a) && !isRequired) () else f(a)
```
actually the reason we didn't factor it was to avoid an un-necessary allocation to `Option`, so I'm gonna ask you to revert (feel free to add a comment to point that out)

astridej replied:
Done, thanks for letting me know!

Baccata commented on modules/bootstrapped/test/src/smithy4s/schema/FieldSpec.scala:
```diff
@@ -52,4 +52,26 @@ final class FieldSpec extends FunSuite {
     assertEquals(result.toList, List(Some("test2"), None))
   }
 
+  test(
```
Thank you for the additional tests. I'd like to see one for the following additional cases : 

```scala
val requiredNullable = string.nullable.required[Foo]("a", _.a)
```

and

```scala
case class Bar(
  a: Option[Nullable[String]]
)  

val optionalNullable = string.nullable.optional[Bar]("a", _.a)
```

astridej replied:
Added! I also took the liberty of adding one for `string.optional` (with no default set), because we didn't have one before and I wanted to make sure that the behaviour I was getting on `string.nullable.optional` was the same as the one on `string.optional`.

Baccata replied:
Awesome, thank you very much !

Baccata commented:
> Digging into it, it seems to come down to two things: [...] 
> If I'm off-base here, let me know!

I think you may be right, but I'd love for the assumption to be validated with a couple additional [test cases](https://github.com/disneystreaming/smithy4s/pull/1408#discussion_r1497774667).

astridej committed:
Revert refactor of getUnlessDefault

astridej committed:
Additional tests for getUnlessDefault

astridej committed:
Remove comment because it's probably misleading

astridej committed:
Add support for Nullable in dynamic module

Including tests to make sure all traits and schemas behave as expected -
this does unfortunately require trying to figure out whether a schema is
Nullable at runtime, which is an awkward process.

kubukoz commented:
I love the emojis in the checklist. Going to steal this for my own :P

astridej committed:
Scalafmt

astridej committed:
Fix trait reference in dynamic module

Not 100% sure why this compiled locally, but it looks like this should
fix it?

astridej committed:
Revert extra handling of default in dynamic module

It's acceptable for fields with a default value to be wrapped in Option
in the dynamically generated code, and trying to align it with the
static codegen behaviour is causing problems in the compliance tests.
We remove that change and any tests that relied on it.

astridej committed:
Add docs for @nullable

astridej committed:
Scalafmt, redux

Apparently 2.13 is more picky about line length for scalafmt than the
other Scala versions, and I hadn't formatted the test files right - this
should fix it.

astridej committed:
Fix wrong copyright on new header

astridej committed:
Update changelog to include nullable

astridej committed:
Correction to the documented code

Baccata commented on modules/docs/markdown/04-codegen/01-customisation/13-nullable-values.md:
```diff
@@ -0,0 +1,56 @@
+title: Nullable Values
+---
+
+By default, smithy does not distinguish between a value being absent and it being null - both are translated to `scala.None`, which in turn will be serialised as an absent value.
```
```suggestion
The official smithy toolset does not offer anything to distinguish between an absence of value and a value set to null during (de) serialisation. 
```

astridej committed:
Update modules/docs/markdown/04-codegen/01-customisation/13-nullable-values.md

Co-authored-by: Olivier Mélois <baccata64@gmail.com>

Baccata reviewed (approved):
This is an amazing PR, both from the added value but also from the fact that it has not required much oversight, and also because it adds some pretty thorough tests. 

I am extremely thankful for this 🥇 

Before we merge, I'd appreciate if @kubukoz or @lewisjkl could offer a second pair of eyes for a review, considering it's a fairly big PR.

astridej commented:
@Baccata thank you for the kind words! 😊 🥳 I'm glad the PR worked without major changes being needed myself - and that you appreciate my edge-case-driven testing approach :)

And thank you for your quick review and good guidance!

lewisjkl reviewed (approved):
This is great work! Thanks for including docs and updating existing docs too. Really appreciate all of it 🙏
